### ISSUE TITLE
Returns 0 IPs

### ISSUE THREAD

**Issue by dominicholt** [jas-/node-libnmap#29]
I am attempting to run a discover function on the network to see all my neighbors using the CIDR, but each time there are zero hosts returned (when I run the same command in nmap itself, 40 hosts are returned). Here is the output:

{"item":{"scanner":"nmap","args":"nmap --host-timeout=900s -oX - -sn --disable-arp-ping fe80::62f8:1dff:fed2:2cd4/ffff:ffff:ffff:ffff::","start":"1450480373","startstr":"Fri Dec 18 15:12:53 2015","version":"7.01","xmloutputversion":"1.04"},"verbose":[{"item":{"level":"0"}}],"debugging":[{"item":{"level":"0"}}],"target":[{"item":{"specification":"fe80::62f8:1dff:fed2:2cd4/ffff:ffff:ffff:ffff::","status":"skipped","reason":"invalid"}}],"runstats":[{"finished":[{"item":{"time":"1450480373","timestr":"Fri Dec 18 15:12:53 2015","elapsed":"0.00","summary":"Nmap done at Fri Dec 18 15:12:53 2015; 0 IP addresses (0 hosts up) scanned in 0.00 seconds","exit":"success"}}],"hosts":[{"item":{"up":"0","down":"0","total":"0"}}]}]}
{"item":{"scanner":"nmap","args":"nmap --host-timeout=900s -oX - -sn --disable-arp-ping fe80::64fd:b3ff:fe67:8c28/ffff:ffff:ffff:ffff::","start":"1450480373","startstr":"Fri Dec 18 15:12:53 2015","version":"7.01","xmloutputversion":"1.04"},"verbose":[{"item":{"level":"0"}}],"debugging":[{"item":{"level":"0"}}],"target":[{"item":{"specification":"fe80::64fd:b3ff:fe67:8c28/ffff:ffff:ffff:ffff::","status":"skipped","reason":"invalid"}}],"runstats":[{"finished":[{"item":{"time":"1450480373","timestr":"Fri Dec 18 15:12:53 2015","elapsed":"0.00","summary":"Nmap done at Fri Dec 18 15:12:53 2015; 0 IP addresses (0 hosts up) scanned in 0.00 seconds","exit":"success"}}],"hosts":[{"item":{"up":"0","down":"0","total":"0"}}]}]}

**Comment by jas-** [comment]
Can you provide your test case code? If you look at the results nmap is flagging your IPv6 range as invalid. Try using verbosity in you options as well.

**Comment by dominicholt** [comment]
Thanks for responding so quickly. My code is very simple:
```
var nmap = require('libnmap')
    , opts = {
    timeout: 900, // 900s = 10m and increases the reliability of scan results
    range: [
        '10.3.3.1/24'
    ]
};

nmap.discover( opts, function(err, report) {
    if (err) throw new Error(err);

    for (var item in report) {
        console.log(JSON.stringify(report[item]));
    }
});
```

**Comment by jas-** [comment]
The range you provided in your example does not accurately reflect the output in your original question

**Comment by dominicholt** [comment]
So, apologies for that, you were totally right, it was a build problem which I've fixed. The true output is: 
```
node-libnmap.js:253
        , ip = new v6.Address(block)
               ^

TypeError: v6.Address is not a function
```
Once again, I apologize if this seems like a simple problem to fix, but I'm very new to Node development. I looked into the actual node-libnmap code where this is referenced and then looked up the ip-address Node project. It appears they have changed the instantiation of IPV6 addresses to "Address6" rather than v6.Address. Since the package.json file grabs the latest version of all the libraries it uses, is this perhaps the problem or is it just how I'm specifying the IP address in my code? Thanks so much for any help.

**Comment by dominicholt** [comment]
As I suspected, after changing line 16 to:
```
, Address6 = require('ip-address').Address6
```
and modifying line 253 to:
```
, ip = new Address6(block)
```
My code compiles and gives this output again:
{"$":{"scanner":"nmap","args":"nmap --host-timeout=900s -oX - -sn --disable-arp-ping fe80::c0cc:ff:feea:51bd/ffff:ffff:ffff:ffff::","start":"1450740857","startstr":"Mon Dec 21 15:34:17 2015","version":"7.01","xmloutputversion":"1.04"},"verbose":[{"$":{"level":"0"}}],"debugging":[{"$":{"level":"0"}}],"target":[{"$":{"specification":"fe80::c0cc:ff:feea:51bd/ffff:ffff:ffff:ffff::","status":"skipped","reason":"invalid"}}],"runstats":[{"finished":[{"$":{"time":"1450740857","timestr":"Mon Dec 21 15:34:17 2015","elapsed":"0.01","summary":"Nmap done at Mon Dec 21 15:34:17 2015; 0 IP addresses (0 hosts up) scanned in 0.01 seconds","exit":"success"}}],"hosts":[{"$":{"up":"0","down":"0","total":"0"}}]}]}
{"$":{"scanner":"nmap","args":"nmap --host-timeout=900s -oX - -sn --disable-arp-ping fe80::62f8:1dff:fed2:2cd4/ffff:ffff:ffff:ffff::","start":"1450740857","startstr":"Mon Dec 21 15:34:17 2015","version":"7.01","xmloutputversion":"1.04"},"verbose":[{"$":{"level":"0"}}],"debugging":[{"$":{"level":"0"}}],"target":[{"$":{"specification":"fe80::62f8:1dff:fed2:2cd4/ffff:ffff:ffff:ffff::","status":"skipped","reason":"invalid"}}],"runstats":[{"finished":[{"$":{"time":"1450740857","timestr":"Mon Dec 21 15:34:17 2015","elapsed":"0.01","summary":"Nmap done at Mon Dec 21 15:34:17 2015; 0 IP addresses (0 hosts up) scanned in 0.01 seconds","exit":"success"}}],"hosts":[{"$":{"up":"0","down":"0","total":"0"}}]}]}

**Comment by jas-** [comment]
Try using 'npm install libnmap' and a valid IPv6 address. Nmap and other cidr IPv6 calculators is flagging 'fe80::62f8:1dff:fed2:2cd4/ffff:ffff:ffff:ffff::' as invalid

**Comment by dominicholt** [comment]
So I installed libnmap, I'm currently using a CIDR address which is valid (as it works in Nmap itself). I've also tried converting it into an IPv6 address and it still says the same thing (and also attempts to change it). Is it possible that the conversion to IPv6 is going wrong somewhere?

**Comment by jas-** [comment]
There is nothing converting or breaking up the IPv6 cidr ranges into smaller blocks like IPv4 cidr ranges. Do the provided test fail as well?

**Comment by dominicholt** [comment]
The tests pass in the libnmap folder.

**Comment by dominicholt** [comment]
So the issue I'm running into is it declares everything as invalid. Doesn't matter if I use a valid IPv4 or IPv6 address, it still skips everything and declares it as invalid. If I use an IPv4 address, it will convert it to IPv6. If I use an IPv6 address, it attempts to convert it anyway (and apparently to an invalid IPv6 address). Perhaps it is because of the way I'm specifying the range for the discover function inside the options above. Is there a more recommended way to do so?

**Comment by jas-** [comment]
The discover method does not require arguments and uses the local adapter(s) address and subnet to scan for neighboring devices.

**Comment by dominicholt** [comment]
If I just plug in the example code on the front page of the project, this is the error I get:

Error: Error: Range must be an array of host(s). Examples: 192.168.2.10 (single), 10.0.2.0/24 (CIDR), 10.0.10.5-20 (range)

**Comment by dominicholt** [comment]
Based on your input of using the local adapters, I wrote this snippet of code:
```
nmap.discover(function(err, report) {
    for (var adapter in ifaces) {
        console.log("Interface xxxxxxxx \n" + JSON.stringify(ifaces[adapter]));

        for( var i = 0 ; i < ifaces[adapter].length ; i++ ) {
            console.log(JSON.stringify(ifaces[adapter][i]));
        }
    }
});
```

Unfortunately this only prints out localhost (and not in the format I'm expecting with the JSON tree that has neighbors). On my network, I know this is not accurate as running a scan with nmap itself yields many hosts.

**Comment by jas-** [comment]
Can you provide the output of `console.log(require('os').networkInterfaces())`. I believe your problem is that you do not have an IPv4 address associated with your adapter indicating a bug in the IPv6 to subnet masking conversion, but I need to confirm it.

**Comment by dominicholt** [comment]
Here is the output you requested:
```
{ lo0: 
   [ { address: '::1',
       netmask: 'ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff',
       family: 'IPv6',
       mac: '00:00:00:00:00:00',
       scopeid: 0,
       internal: true },
     { address: '127.0.0.1',
       netmask: '255.0.0.0',
       family: 'IPv4',
       mac: '00:00:00:00:00:00',
       internal: true },
     { address: 'fe80::1',
       netmask: 'ffff:ffff:ffff:ffff::',
       family: 'IPv6',
       mac: '00:00:00:00:00:00',
       scopeid: 1,
       internal: true } ],
  en0: 
   [ { address: 'fe80::62f8:1dff:fed2:2cd4',
       netmask: 'ffff:ffff:ffff:ffff::',
       family: 'IPv6',
       mac: '60:f8:1d:d2:2c:d4',
       scopeid: 4,
       internal: false },
     { address: '10.3.3.229',
       netmask: '255.255.255.0',
       family: 'IPv4',
       mac: '60:f8:1d:d2:2c:d4',
       internal: false } ],
  awdl0: 
   [ { address: 'fe80::2c9b:a4ff:fefb:99fb',
       netmask: 'ffff:ffff:ffff:ffff::',
       family: 'IPv6',
       mac: '2e:9b:a4:fb:99:fb',
       scopeid: 8,
       internal: false } ] }
```
I've upgraded to v0.2.23 but still seem to be having the same issues.

**Comment by jas-** [comment]
Is this a production system? Are you able to offline the `awd10` adapter and try it? I am fairly certain this is a bug with the `IPv6` adapter. The `discover` mode attempts to generate a `CIDR` range from the `addr` and `subnet` associated with `awd10` and I don't believe `ffff:ffff:ffff:ffff::` is a valid notation.

**Comment by jas-** [comment]
@dominicholt Is this still an issue?

**Comment by dominicholt** [comment]
Sorry there's been radio silence over here, I haven't been able to work on this project much due to other obligations. I just updated and still only seem to get 1 host returned. What can I provide to help?

**Comment by jas-** [comment]
Can you turn on verbose mode and post the results?

**Comment by dominicholt** [comment]
Network Interfaces Output: 

{ lo0: 
   [ { address: '::1',
       netmask: 'ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff',
       family: 'IPv6',
       mac: '00:00:00:00:00:00',
       scopeid: 0,
       internal: true },
     { address: '127.0.0.1',
       netmask: '255.0.0.0',
       family: 'IPv4',
       mac: '00:00:00:00:00:00',
       internal: true },
     { address: 'fe80::1',
       netmask: 'ffff:ffff:ffff:ffff::',
       family: 'IPv6',
       mac: '00:00:00:00:00:00',
       scopeid: 1,
       internal: true } ],
  en0: 
   [ { address: 'fe80::62f8:1dff:fed2:2cd4',
       netmask: 'ffff:ffff:ffff:ffff::',
       family: 'IPv6',
       mac: '60:f8:1d:d2:2c:d4',
       scopeid: 4,
       internal: false },
     { address: '10.3.3.23',
       netmask: '255.255.255.0',
       family: 'IPv4',
       mac: '60:f8:1d:d2:2c:d4',
       internal: false } ],
  awdl0: 
   [ { address: 'fe80::1c00:2bff:fecc:9d9',
       netmask: 'ffff:ffff:ffff:ffff::',
       family: 'IPv6',
       mac: '1e:00:2b:cc:09:d9',
       scopeid: 8,
       internal: false } ] }
Running: nmap  --host-timeout=120s -oX - -sn --disable-arp-ping -6 fe80::62f8:1dff:fed2:2cd4/128
Running: nmap  --host-timeout=120s -oX - -sn --disable-arp-ping -6 fe80::1c00:2bff:fecc:9d9/128
Interface xxxxxxxx 
[{"address":"::1","netmask":"ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff","family":"IPv6","mac":"00:00:00:00:00:00","scopeid":0,"internal":true},{"address":"127.0.0.1","netmask":"255.0.0.0","family":"IPv4","mac":"00:00:00:00:00:00","internal":true},{"address":"fe80::1","netmask":"ffff:ffff:ffff:ffff::","family":"IPv6","mac":"00:00:00:00:00:00","scopeid":1,"internal":true}]
{"address":"::1","netmask":"ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff","family":"IPv6","mac":"00:00:00:00:00:00","scopeid":0,"internal":true}
{"address":"127.0.0.1","netmask":"255.0.0.0","family":"IPv4","mac":"00:00:00:00:00:00","internal":true}
{"address":"fe80::1","netmask":"ffff:ffff:ffff:ffff::","family":"IPv6","mac":"00:00:00:00:00:00","scopeid":1,"internal":true}
Interface xxxxxxxx 
[{"address":"fe80::62f8:1dff:fed2:2cd4","netmask":"ffff:ffff:ffff:ffff::","family":"IPv6","mac":"60:f8:1d:d2:2c:d4","scopeid":4,"internal":false},{"address":"10.3.3.23","netmask":"255.255.255.0","family":"IPv4","mac":"60:f8:1d:d2:2c:d4","internal":false}]
{"address":"fe80::62f8:1dff:fed2:2cd4","netmask":"ffff:ffff:ffff:ffff::","family":"IPv6","mac":"60:f8:1d:d2:2c:d4","scopeid":4,"internal":false}
{"address":"10.3.3.23","netmask":"255.255.255.0","family":"IPv4","mac":"60:f8:1d:d2:2c:d4","internal":false}
Interface xxxxxxxx 
[{"address":"fe80::1c00:2bff:fecc:9d9","netmask":"ffff:ffff:ffff:ffff::","family":"IPv6","mac":"1e:00:2b:cc:09:d9","scopeid":8,"internal":false}]
{"address":"fe80::1c00:2bff:fecc:9d9","netmask":"ffff:ffff:ffff:ffff::","family":"IPv6","mac":"1e:00:2b:cc:09:d9","scopeid":8,"internal":false}
{"item":{"scanner":"nmap","args":"nmap --host-timeout=120s -oX - -sn --disable-arp-ping -6 fe80::62f8:1dff:fed2:2cd4/128","start":"1454522817","startstr":"Wed Feb  3 10:06:57 2016","version":"7.01","xmloutputversion":"1.04"},"verbose":[{"item":{"level":"0"}}],"debugging":[{"item":{"level":"0"}}],"host":[{"status":[{"item":{"state":"up","reason":"syn-ack","reason_ttl":"0"}}],"address":[{"item":{"addr":"fe80::62f8:1dff:fed2:2cd4","addrtype":"ipv6"}}],"hostnames":["\n"],"times":[{"item":{"srtt":"4743","rttvar":"5000","to":"100000"}}]}],"runstats":[{"finished":[{"item":{"time":"1454522817","timestr":"Wed Feb  3 10:06:57 2016","elapsed":"0.08","summary":"Nmap done at Wed Feb  3 10:06:57 2016; 1 IP address (1 host up) scanned in 0.08 seconds","exit":"success"}}],"hosts":[{"item":{"up":"1","down":"0","total":"1"}}]}]}
{"item":{"scanner":"nmap","args":"nmap --host-timeout=120s -oX - -sn --disable-arp-ping -6 fe80::1c00:2bff:fecc:9d9/128","start":"1454522817","startstr":"Wed Feb  3 10:06:57 2016","version":"7.01","xmloutputversion":"1.04"},"verbose":[{"item":{"level":"0"}}],"debugging":[{"item":{"level":"0"}}],"host":[{"status":[{"item":{"state":"up","reason":"syn-ack","reason_ttl":"0"}}],"address":[{"item":{"addr":"fe80::1c00:2bff:fecc:9d9","addrtype":"ipv6"}}],"hostnames":["\n"],"times":[{"item":{"srtt":"186","rttvar":"5000","to":"100000"}}]}],"runstats":[{"finished":[{"item":{"time":"1454522817","timestr":"Wed Feb  3 10:06:57 2016","elapsed":"0.26","summary":"Nmap done at Wed Feb  3 10:06:57 2016; 1 IP address (1 host up) scanned in 0.26 seconds","exit":"success"}}],"hosts":[{"item":{"up":"1","down":"0","total":"1"}}]}]}

**Comment by jas-** [comment]
Both `fe80::62f8:1dff:fed2:2cd4/128` & `fe80::1c00:2bff:fecc:9d9/128` are single IP ranges.

I may have to address this according to the nmap documentation regarding [host discovery](https://nmap.org/book/man-host-discovery.html)

> -PR (ARP Ping)

    One of the most common Nmap usage scenarios is to scan an ethernet LAN. On most LANs, especially those using private address ranges specified by RFC 1918, the vast majority of IP addresses are unused at any given time. When Nmap tries to send a raw IP packet such as an ICMP echo request, the operating system must determine the destination hardware (ARP) address corresponding to the target IP so that it can properly address the ethernet frame. This is often slow and problematic, since operating systems weren't written with the expectation that they would need to do millions of ARP requests against unavailable hosts in a short time period.

    ARP scan puts Nmap and its optimized algorithms in charge of ARP requests. And if it gets a response back, Nmap doesn't even need to worry about the IP-based ping packets since it already knows the host is up. This makes ARP scan much faster and more reliable than IP-based scans. So it is done by default when scanning ethernet hosts that Nmap detects are on a local ethernet network. Even if different ping types (such as -PE or -PS) are specified, Nmap uses ARP instead for any of the targets which are on the same LAN. If you absolutely don't want to do an ARP scan, specify --disable-arp-ping.

    For IPv6 (-6 option), -PR uses ICMPv6 Neighbor Discovery instead of ARP. Neighbor Discovery, defined in RFC 4861, can be seen as the IPv6 equivalent of ARP.

**Comment by jas-** [comment]
If you get a minute could you try the `v0.2.24` branch from here vs. NPM? I don't have an environment for IPv6 validation.

**Comment by cscholze** [comment]
I have been having a similar issue with the discovery scan.  I'm using libnmap with OSX.  From what I've discovered, libnmap uses node's 'os' module to grab the network interfaces in the 'adapter' function.  When looping through the adapters object, it grabs the first item in the iface array , which corresponds with the IPv6 address for the adapters.  I modified the code to grab the last item, which now grabs the IPv6 address for the awdl0 interface, and the IPv4 address for en0.  The discovery scan then passes both of these to the discovery scan as ranges and now includes IPv4 hosts in the /24 subnet on my local network.  The IPv6 address acquired from the awdl0 interface still fails to return any hosts during the scan.  I also removed the -PR default discovery scan flag and added --disable-arp-ping.

I will try to update the code to include all IPv4 and IPv6 addresses from the network interfaces to make it more robust.
